### ISSUE TITLE
Customizing Frame Size for Tree Node Headers?

### ISSUE THREAD

**Issue by satasta** [ocornut/imgui#7139]
Adjusting the size of the frame around a tree node using ImGuITreeNodeFlags_framed isn't an option that's directly supported for configuration??
How can I modify the width (by .x) of the frame's header?
**I'd prefer the frame not to extend across the entire window.**
 Are there any solutions or workarounds available?

**Comment by ocornut** [comment]
Hello,

You'd need to provide more context and intent if you want an answer.

**Comment by satasta** [comment]
For instance, I aim to shorten the tree frame (the blue frame) from its default length. How can I adjust the size of this blue frame to my preferred value? The red lines illustrate the desired size (some value...)

![image](https://github.com/ocornut/imgui/assets/154091983/db920173-50de-463b-8371-8ae5a5d26b9d)

**Comment by ocornut** [comment]
Hello,

The widget doesn't honor `SetNextItemWidth()` because it would be inconsistent if it did so and didn't honor the `PushItemWidth()`-stack which is "always" set to something and not defaulting to right-alignment.

One planned change is that the yet-unexisting `PushWorkRect()` would reprogram this.
The easier workaround I would suggest now would be to temporarily write to `g.CurrentWindow->WorkRect.Max.x`, call the function then restore original value. You may also need to temporarily set `style.WindowPadding.x` to 0.0f to have an exact value (but that part will not be needed once we make the change mentioned in this comment: https://github.com/ocornut/imgui/issues/3312#issuecomment-648052521)

The more general thing is that we have quite some tension with the outer API or both TreeNode() and Selectable() as part as their feature set which has quite some overlap and there's a task to rework this more toroughly later.

**Comment by ocornut** [comment]
Here's the code for suggested workaround:

```cpp
bool CollapsingHeaderWithWidth(const char* label, ImGuiTreeNodeFlags flags = 0, float width = 0.0f)
{
    if (width == 0.0f)
        return ImGui::CollapsingHeader(label, flags);

    ImGuiContext& g = *GImGui;
    ImGuiWindow* window = g.CurrentWindow;
    ImGui::SetNextItemWidth(width);
    float backup_work_max_x = window->WorkRect.Max.x;
    window->WorkRect.Max.x = window->DC.CursorPos.x + ImGui::CalcItemWidth();
    bool ret = ImGui::CollapsingHeader(label, flags);
    window->WorkRect.Max.x = backup_work_max_x;
    return ret;
}
```

Usage:
```cpp
if (CollapsingHeaderWithWidth("Fixed width", 0, 200.0f))
{
    ImGui::Text("Hiii");
}

if (CollapsingHeaderWithWidth("Right aligned", 0, -50.0f))
{
    ImGui::Text("Hiii");
}
```
